document.go

```diff
@@ -186,3 +241,49 @@ func (d *Document) Error() string {
 	}
 	return errStr
 }
+
+/*
+MarshalJSON handles the custom serialization case caused by case where the "data"
+element of a document might be either a single resource object, or a collection of
+them.
+*/
+func (d *Document) MarshalJSON() ([]byte, error) {
+	type MarshalDoc Document
+	doc := MarshalDoc(*d)
+
+	// get raw data so we can perform top level operations
+	rawData, err := json.Marshal(doc)
+	if err != nil {
+		return nil, err
+	}
+
+	// map back to top level JSON map for ease of use, SHITTY, I KNOW.
+	jDoc := map[string]interface{}{}
```

Comment:
@petercgrant if you have any suggestions here on how to achieve this in a saner way, I am open to suggestions.